### PR TITLE
feat(state_store): cut range to make sure that they are in order

### PR DESCRIPTION
**Pull request by Little-Wallace**
Signed-off-by: Little-Wallace <bupt2013211450@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

To support filter keys which deleted by range-tombstone, we would construct a special binary-heap in  https://github.com/risingwavelabs/risingwave/pull/6279
However, if there are a lot of range-tombstones, we must to check them one by one to make sure that the current key and following keys would never overlap with these range because we only sort these range-tombstones by `start_user_key` rather than `end_user_key`.

So to reduce the effect to query( in other words, to optimize the performance of delete-range-filter) , we must make sure that both  `start_user_key`  and `end_user_key` of tombstones are all in order. 

I would cut range which overlap with each other to ensure this order. For example:

* [a, c) and [b, d) are legal, because a > b while b > d.
* [a, d) and [b, c)  are illegal, I would split them into three ranges: [a, b), [b, c) and [c, d).

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

https://github.com/risingwavelabs/risingwave/issues/5978

### CODE REVIEW

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs [hunk 1]
```diff
@@ -48,20 +48,35 @@ pub(crate) struct SharedBufferBatchInner {
...
+                // and does not affect correctness.
+                if largest_user_key.lt(&tombstone.end_user_key) {
+                    largest_user_key.clear();
+                    largest_user_key.extend_from_slice(&tombstone.end_user_key);
+                }
+                if let Some(last) = range_tombstones.last_mut() {
+                    if last.end_user_key.gt(&tombstone.start_user_key) {
+                        if last.end_user_key.lt(&tombstone.end_user_key) {
+                            last.end_user_key = tombstone.end_user_key;
+                        }
```
Comment (soundOfDestiny) [inline src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:71]: again, why can we omit a tombstone with unknown sequence?

File: src/storage/src/hummock/sstable/mod.rs [hunk 2]
```diff
@@ -64,6 +64,26 @@ pub struct DeleteRangeTombstone {
     pub sequence: HummockEpoch,
 }
 
+impl PartialOrd for DeleteRangeTombstone {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
```
Comment (soundOfDestiny) [inline src/storage/src/hummock/sstable/mod.rs:68]: Can we use Some(cmp()) here?

File: src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs (hunk 1 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs:71]: No. The sequence must be right. Here we assign sequence for new tombstone by the origin tombstone.

File: src/storage/src/hummock/sstable/mod.rs (hunk 2 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/mod.rs:68]: Yes. I would change it

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs [hunk 3]
```diff
@@ -124,8 +156,19 @@ impl RangeTombstonesCollector {
                 continue;
             }
 
-            if self.gc_delete_keys && tombstone.sequence <= self.watermark {
-                continue;
+            if tombstone.sequence <= self.watermark {
+                if self.gc_delete_keys {
```
Comment (soundOfDestiny) [inline src/storage/src/hummock/sstable/delete_range_aggregator.rs:160]: why?

**Review by soundOfDestiny — APPROVED** [review]
rest lgtm

File: src/storage/src/hummock/sstable/mod.rs (hunk 2 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/mod.rs:71]: done

File: src/storage/src/hummock/sstable/delete_range_aggregator.rs (hunk 3 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/delete_range_aggregator.rs:157]: We can drop this range-tombstone when it reach the bottommost level just like other normal delete-tombstone.

**Pull request merged**